### PR TITLE
Add concurrentAsyncExpand

### LINKED ISSUE #77: Add concurrentAsyncExpand?
We have `Stream.asyncMap`, `concurrentAsyncMap`, `Stream.asyncExpand`... but no `concurrentAsyncExpand`.

```dart
StreamTransformer<S, T> concurrentAsyncExpand<S, T>(Stream<T> convert(S event))
```

Example:

```dart
Iterable<Future<int>> futures = [3, 2, 1].map((i) => Future.delayed(Duration(seconds: i), () => i));
Stream.fromIterable(futures)
  .transform(concurrentAsyncExpand((future) async* {
    yield 'awaiting future';
    var result = await future;
    yield 'future completed: $result';
  }))
  .forEach(print);
```

Output:
```
awaiting future
awaiting future
awaiting future
future completed: 1
future completed: 2
future completed: 3
```

### LINKED ISSUE #1765: Add concurrentAsyncExpand?
We have `Stream.asyncMap`, `concurrentAsyncMap`, `Stream.asyncExpand`... but no `concurrentAsyncExpand`.

```dart
StreamTransformer<S, T> concurrentAsyncExpand<S, T>(Stream<T> convert(S event))
```

Example:

```dart
Iterable<Future<int>> futures = [3, 2, 1].map((i) => Future.delayed(Duration(seconds: i), () => i));
Stream.fromIterable(futures)
  .transform(concurrentAsyncExpand((future) async* {
    yield 'awaiting future';
    var result = await future;
    yield 'future completed: $result';
  }))
  .forEach(print);
```

Output:
```
awaiting future
awaiting future
awaiting future
future completed: 1
future completed: 2
future completed: 3
```

### PR DESCRIPTION
**Pull request by natebosch**
Closes dart-lang/tools#1765

Added with the `Merge` extension, since the behavior is similar to
merging streams that are created lazily.

Adds a private `StreamTransformer` to merge a `Stream<Stream<T>>`. This
could be exposed separately as an extension on type.

In order to allow cancelling and then re-listening on broadcast streams,
an sub stream which is single subscription must be converted to
broadcast. This can be risky since subscriptions can leak and cause the
inner streams to never be canceled.

### CODE REVIEW

File: lib/src/merge.dart [hunk 1]
```diff
@@ -165,12 +165,12 @@ class _MergeExpanded<T> extends StreamTransformerBase<Stream<T>, T> {
       if (!streams.isBroadcast) {
         controller
           ..onPause = () {
-            for (var subscription in subscriptions) {
+            for (final subscription in subscriptions) {
```
Comment (grouma) [inline lib/src/merge.dart:168]: Why this change?

**Review by grouma — APPROVED** [review]

File: lib/src/merge.dart (hunk 1 above)
Comment (natebosch) [reply lib/src/merge.dart:168]: It matches what I did in https://github.com/dart-lang/stream_transform/pull/92

As I have been rereading this code I found it easier to have a stronger signal for which variables were reassigned so I started replacing more `var` with `final`. It's less of a big deal in this small for loop, but I'm strongly considering turning on the lint to enforce it for this package.
